readme.md

```diff
@@ -47,126 +62,497 @@ This package exports [ESM](https://gist.github.com/sindresorhus/a39789f98801d908
+  - Creates a new instance of the model with modified configuration
 
-You'll need the following environment variables to run this example:
+### Extract Functions
 
-- `OPENAI_API_KEY`
-- `PINECONE_API_KEY`
-- `PINECONE_BASE_URL`
-  - Note: Pinecone's free starter indexes don't seem to support namespaces or hybrid search, so unfortunately you'll need to upgrade to a paid plan to run this example. See Pinecone's [hybrid docs](https://docs.pinecone.io/docs/hybrid-search) for details on setting up a hybrid index, and make sure it is using the `dotproduct` metric.
-- `SPLADE_SERVICE_URL`
-  - Here is an [example](https://gist.github.com/transitive-bullshit/cc9140ff832fc7e815a48f0a45e1fc27) of how to run a SPLADE REST API, which can be deployed to [Modal](https://modal.com) or any other GPU-enabled hosting provider.
+#### createExtractFunction
 
-```bash
-npx tsx examples/chatbot/ingest.ts
+Creates a function to extract structured data from text using OpenAI's structured output feature.
+This is a better way to extract structured data than using the legacy `createAIExtractFunction` function.
+
+```typescript
+createExtractFunction<Schema extends z.ZodObject<any>>(args: {
+  chatModel: Model.Chat.Model;
+  name: string;
+  schema: Schema;
+  systemMessage: string;
+}): (input: string | Msg) => Promise<z.infer<Schema>>
 ```
 
-```bash
-npx tsx examples/chatbot/cli.ts
+### AI Functions
+
+#### createAIFunction
+
+Creates a function meant to be used with OpenAI tool or function calling.
+
+```typescript
+createAIFunction<Schema extends z.ZodObject<any>, Return>(
+  spec: {
+    name: string;
+    description?: string;
+    argsSchema: Schema;
+  },
+  implementation: (params: z.infer<Schema>) => Promise<Return>
+): AIFunction<Schema, Return>
+```
+
+#### createAIExtractFunction
+
+Creates a function to extract structured data from text using OpenAI function calling.
+
+```typescript
+createAIExtractFunction<Schema extends z.ZodObject<any>>(
+  {
+    chatModel: Model.Chat.Model;
+    name: string;
+    description?: string;
+    schema: Schema;
+    maxRetries?: number;
+    systemMessage?: string;
+    functionCallConcurrency?: number;
+  },
+  customExtractImplementation?: (params: z.infer<Schema>) => z.infer<Schema> | Promise<z.infer<Schema>>
+): ExtractFunction<Schema>
 ```
 
-[source](./examples/chatbot)
+#### createAIRunner
+
+Creates a function to run a chat model in a loop, handling parsing, running, and inserting responses for function & tool call messages.
+
+```typescript
+createAIRunner<Content = string>(args: {
+  chatModel: Model.Chat.Model;
+  functions?: AIFunction[];
+  shouldBreakLoop?: (msg: Msg) => boolean;
+  maxIterations?: number;
+  functionCallConcurrency?: number;
+  validateContent?: (content: string | null) => Content | Promise<Content>;
+  mode?: Runner.Mode;
+  systemMessage?: string;
+  onRetriableError?: (error: Error) => void;
+}): Runner<Content>
+```
+
+### Utilities
+
+#### MsgUtil
```

Comment:
Not sure I like `MsgUtil`. What was wrong with `Msg`? Or `Message`?

Reply:
`Msg` is the type and `Message` would be confusing with the `Msg` type.